### PR TITLE
Seperate HIR owner from LocalDefId in the type system

### LINKED ISSUE #83158: Separate definitions and HIR owners in the type system
Both definitions and HIR owners are represented by a LocalDefId,
but they don't have the same semantics.

HIR owners are the `owner` field of a `HirId`,
and the argument passed to the `hir_owner*` queries.
HIR owners contain the following HIR nodes: Item, TraitItem, ImplItem, ForeignItem, MacroDef,
and `GenericParam`s which come from `impl Trait` desugaring.

LocalDefId enumerate definitions in the local crate.
Definitions are a broader concept than HIR owners.
For instance, closures are definitions but not HIR owners.

On the other hand, all HIR owners are definitions.

We need to introduce a new type `HirOwner` type (name to bikeshed),
whose usage will be the following:
- `HirOwner` is defined as `struct HirOwner { def_id: LocalDefId }`;
- `HirId` becomes `{ owner: HirOwner, local_id: ItemLocalId }`;
- Item, TraitItem, ImplItem, ForeignItem, MacroDef have a `HirOwner` as their `def_id` field;
- `hir_owner*` queries take a `HirOwner` as argument;
- the passage between `HirOwner` and `LocalDefId` happens through
  the `local_def_id_to_hir_id: LocalDefId -> HirId`
  and `local_def_id: HirId -> LocalDefId` methods.

I am available on Zulip for further information.

### PR DESCRIPTION
**Pull request by ABouttefeux**
closes #83158

Separate definitions and HIR owners in the type system

### CODE REVIEW

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -2890,7 +2890,7 @@ pub struct ForeignItemRef<'hir> {
 pub struct ForeignItem<'hir> {
     pub ident: Ident,
     pub kind: ForeignItemKind<'hir>,
-    pub def_id: LocalDefId,
+    pub def_id: HirOwner,
```
Comment (bjorn3) [inline compiler/rustc_hir/src/hir.rs:2893]: I think `owner` would be a better name now for this field here and elsewhere.

File: src/tools/clippy/clippy_utils/src/usage.rs [hunk 2]
```diff
@@ -189,7 +194,7 @@ impl<'tcx> Visitor<'tcx> for ReturnBreakContinueMacroVisitor {
                 } else {
                     rustc_hir::intravisit::walk_expr(self, ex);
                 }
-            },
+            }
```
Comment (bjorn3) [inline src/tools/clippy/clippy_utils/src/usage.rs:197]: Can you please revert all those formatting changes?

Comment (ABouttefeux) [reply src/tools/clippy/clippy_utils/src/usage.rs:197]: I didn't notice I am reverting it.

File: compiler/rustc_hir/src/hir.rs (hunk 1 above)
Comment (ABouttefeux) [reply compiler/rustc_hir/src/hir.rs:2893]: There is already a struct Owner here https://github.com/rust-lang/rust/blob/fbf1b1a7193cda17008ab590e06ad28d9924023b/compiler/rustc_middle/src/hir/mod.rs#L34-L37

**Comment by crlf0710** [comment]
@ABouttefeux Ping from triage. There's merge conflict now.

File: compiler/rustc_middle/src/ty/context.rs [hunk 3]
```diff
@@ -2663,26 +2664,26 @@ impl<'tcx> TyCtxt<'tcx> {
     }
 
     pub fn in_scope_traits(self, id: HirId) -> Option<&'tcx StableVec<TraitCandidate>> {
-        self.in_scope_traits_map(id.owner).and_then(|map| map.get(&id.local_id))
+        self.in_scope_traits_map(id.owner.def_id).and_then(|map| map.get(&id.local_id))
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:2667]: Can `in_scope_traits_map` take a HirOwner?

File: compiler/rustc_middle/src/ty/context.rs [hunk 4]
```diff
@@ -2663,26 +2664,26 @@ impl<'tcx> TyCtxt<'tcx> {
...
     }
 
     pub fn named_region(self, id: HirId) -> Option<resolve_lifetime::Region> {
         debug!(?id, "named_region");
-        self.named_region_map(id.owner).and_then(|map| map.get(&id.local_id).cloned())
+        self.named_region_map(id.owner.def_id).and_then(|map| map.get(&id.local_id).cloned())
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:2672]: Can `named_region_map` take a HirOwner?

File: compiler/rustc_middle/src/ty/context.rs [hunk 5]
```diff
@@ -2663,26 +2664,26 @@ impl<'tcx> TyCtxt<'tcx> {
...
     }
 
     pub fn is_late_bound(self, id: HirId) -> bool {
-        self.is_late_bound_map(id.owner)
-            .map_or(false, |(owner, set)| owner == id.owner && set.contains(&id.local_id))
+        self.is_late_bound_map(id.owner.def_id)
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:2676]: Can `late_bound_map` take a HirOwner?

File: compiler/rustc_middle/src/ty/context.rs [hunk 6]
```diff
@@ -2663,26 +2664,26 @@ impl<'tcx> TyCtxt<'tcx> {
...
+            .map_or(false, |(owner, set)| owner == id.owner.def_id && set.contains(&id.local_id))
     }
 
     pub fn object_lifetime_defaults(self, id: HirId) -> Option<Vec<ObjectLifetimeDefault>> {
-        self.object_lifetime_defaults_map(id.owner)
+        self.object_lifetime_defaults_map(id.owner.def_id)
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:2681]: Can `object_lifetime_defaults_map` take a HirOwner?

File: compiler/rustc_middle/src/ty/context.rs [hunk 7]
```diff
@@ -2663,26 +2664,26 @@ impl<'tcx> TyCtxt<'tcx> {
...
     }
 
     pub fn late_bound_vars(self, id: HirId) -> &'tcx List<ty::BoundVariableKind> {
         self.mk_bound_variable_kinds(
-            self.late_bound_vars_map(id.owner)
+            self.late_bound_vars_map(id.owner.def_id)
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:2686]: Can `late_bound_vars_map` take a HirOwner?

File: compiler/rustc_middle/src/ty/context.rs [hunk 8]
```diff
@@ -2692,7 +2693,7 @@ impl<'tcx> TyCtxt<'tcx> {
     }
 
     pub fn lifetime_scope(self, id: HirId) -> Option<LifetimeScopeForPath> {
-        self.lifetime_scope_map(id.owner).and_then(|mut map| map.remove(&id.local_id))
+        self.lifetime_scope_map(id.owner.def_id).and_then(|mut map| map.remove(&id.local_id))
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:2696]: Can `lifetime_scope_map` take a HirOwner?

File: compiler/rustc_hir/src/hir_id.rs [hunk 9]
```diff
@@ -1,7 +1,54 @@
...
+impl HirOwner {
+    pub fn def_id(&self) -> LocalDefId {
+        self.def_id
+    }
+
+    pub fn to_def_id(&self) -> DefId {
+        self.def_id.to_def_id()
+    }
+
+    pub fn to_hir_id(self) -> HirId {
```
Comment (cjgillot) [inline compiler/rustc_hir/src/hir_id.rs:20]: This deserves to be `#[inline]`.

File: compiler/rustc_hir/src/hir_id.rs [hunk 10]
```diff
@@ -1,7 +1,54 @@
...
+pub struct HirOwner {
+    pub def_id: LocalDefId,
+}
+
+impl HirOwner {
+    pub fn def_id(&self) -> LocalDefId {
+        self.def_id
+    }
+
+    pub fn to_def_id(&self) -> DefId {
```
Comment (cjgillot) [inline compiler/rustc_hir/src/hir_id.rs:16]: This deserves to be `#[inline]`.

File: compiler/rustc_hir/src/hir_id.rs [hunk 11]
```diff
@@ -1,7 +1,54 @@
...
+use rustc_serialize::{Decodable, Decoder, Encodable, Encoder};
 use std::fmt;
 
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct HirOwner {
+    pub def_id: LocalDefId,
+}
+
+impl HirOwner {
+    pub fn def_id(&self) -> LocalDefId {
```
Comment (cjgillot) [inline compiler/rustc_hir/src/hir_id.rs:12]: Is this method useful?
It deserves to be `#[inline]`.

File: compiler/rustc_middle/src/ty/query/mod.rs [hunk 12]
```diff
@@ -308,6 +308,13 @@ mod sealed {
             self.to_def_id()
         }
     }
+
+    impl IntoQueryParam<DefId> for HirOwner {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/query/mod.rs:312]: Great idea!
An impl `IntoQueryParam<LocalDefId>` can also help a lot.

File: compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs [hunk 13]
```diff
@@ -698,7 +698,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
             {
                 let opt_suggestions = path_segment
                     .hir_id
-                    .map(|path_hir_id| self.infcx.tcx.typeck(path_hir_id.owner))
+                    .map(|path_hir_id| self.infcx.tcx.typeck(path_hir_id.owner.def_id))
```
Comment (cjgillot) [inline compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs:701]: Can `typeck` take a HirOwner?

File: compiler/rustc_mir/src/borrow_check/universal_regions.rs [hunk 14]
```diff
@@ -804,7 +804,7 @@ fn for_each_late_bound_region_defined_on<'tcx>(
 ) {
     if let Some((owner, late_bounds)) = tcx.is_late_bound_map(fn_def_id.expect_local()) {
```
Comment (cjgillot) [inline compiler/rustc_mir/src/borrow_check/universal_regions.rs:805]: Can this `owner` be a `HirOwner` directly?

**Review by cjgillot — CHANGES_REQUESTED** [review]
Thank you for the PR @ABouttefeux and sorry for the delay!
Overall, this is going in the right direction.
There is still a lot of back-and-forth between HirOwner and LocalDefId. I left a few comment to reduce the conversion glue.

**Comment by ABouttefeux** [comment]
Sorry, currently I am hospitalized and I am  not able to work in this PR.

**Comment by bstrie** [comment]
@ABouttefeux Hope you feel better soon!

**Comment by crlf0710** [comment]
@ABouttefeux Ping from triage! Any updates on this?

**Comment by ABouttefeux** [comment]
I am still in the hospital and unable to work on it. It may take a while before I am able to work on it.

**Comment by JohnCSimon** [comment]
Ping from triage:
 
@ABouttefeux closing this as inactive. Please feel free to reopen when you are ready to continue.

**Pull request closed without merging**